File: ocaml-tiger/line.ml
type id = string

type binop = Plus | Minus | Times | Div

type stm = CompoundStm of stm * stm
         | AssignStm of id * exp
         | PrintStm of exp list
and exp = IdExp of id
        | NumExp of int
        | OpExp of exp * binop * exp
        | EseqExp of stm * exp

let prog =
  CompoundStm(AssignStm("a", OpExp(NumExp 5, Plus, NumExp 3)),
              CompoundStm(AssignStm("b",
                                    EseqExp(PrintStm [IdExp "a"; OpExp(IdExp "a", Minus, NumExp 1)],
                                            OpExp(NumExp 10, Times, IdExp "a"))),
                          PrintStm [IdExp "b"]))

let rec maxargs stm : int =
  match stm with
  | CompoundStm(sl, sr) -> max (maxargs sl) (maxargs sr)
  | AssignStm(_, e) -> maxargsE e
  | PrintStm es -> List.fold_left (fun m s -> max m (maxargsE s)) (List.length es) es
and maxargsE exp : int =
  match exp with
  | EseqExp(s, e) -> max (maxargs s) (maxargsE e)
  | _ -> 0

module type ST = sig
  type t

  val add_sym : id -> int -> t -> t

  val value_of : id -> t -> int

  val append_output : string -> t -> t

  val output : t -> string list
end

module BasicState : ST = struct
  type t = { symbol_table : (id * int) list ; output : string list }

  let add_sym l v st =
    { symbol_table = (l, v) :: List.remove_assoc l st.symbol_table ;
      output = st.output }

  let value_of l st = List.assoc l st.symbol_table

  let append_output l st =
    { symbol_table = st.symbol_table ; output = l :: st.output }

  let output st = st.output
end

module S = BasicState

let runOp op vl vr =
  match op with
  | Plus -> vl + vr
  | Minus -> vl - vr
  | Times -> vl * vr
  | Div -> vl / vr

let rec evalE e (s : S.t) : (S.t * int) =
  match e with
  | IdExp(label) -> (s, S.value_of label s)
  | NumExp(n) -> (s, n)
  | OpExp(el, op, er) ->
    let (s', vl) = evalE el s in
    let (s'', vr) = evalE er s' in
    (s'', runOp op vl vr)
  | EseqExp(stm, exp) ->
    let s' = evalS stm s in
    evalE exp s'
and evalS s st : S.t =
  match s with
  | CompoundStm(sl, sr) ->
    let st' = evalS sl st in
    evalS sr st'
  | AssignStm(label, e) ->
    let (s', v) = evalE e st in
    S.add_sym label v s'
  | PrintStm _ -> st (* how do i print stuff? *)

type key = string

type tree = Leaf
          | Branch of tree * key * tree

let empty = Leaf

(* Exercise 1.1 *)

let rec insert key = function
  | Leaf -> Branch(Leaf, key, Leaf)
  | Branch(lt, k, rt) ->
    if key < k
    then Branch(insert key lt, k, rt)
    else if key > k
    then Branch(lt, k, insert key rt)
    else Branch(lt, k, rt)

let rec member key = function
  | Leaf -> false
  | Branch(lt, k, rt) ->
    if key < k
    then member k lt
    else if key > k
    then member k rt
    else true

type 'a kv_tree = Leaf
                | Branch of 'a kv_tree * key * 'a * 'a kv_tree

let rec kv_insert key value = function
  | Leaf -> Branch(Leaf, key, value, Leaf)
  | Branch(lt, k, v, rt) ->
    if key < k
    then Branch(kv_insert key value lt, k, v, rt)
    else if key > k
    then Branch(lt, k, v, kv_insert key value rt)
    else Branch(lt, k, v, rt)

let rec kv_member key = function
  | Leaf -> false
  | Branch(lt, k, _, rt) ->
    if key < k
    then kv_member k lt
    else if key > k
    then kv_member k rt
    else true

let rec lookup key = function
  | Leaf -> None
  | Branch(lt, k, v, rt) ->
    if key < k
    then lookup key lt
    else if key > k
    then lookup key rt
    else v

File: haskell-tiger/src/Line.hs
module Line where

import Data.List (find)
import Data.Maybe (fromJust)

type Id = String

data BinOp = Plus | Minus | Times | Div deriving (Show, Eq)

data Exp = IdExp Id
         | NumExp Int
         | OpExp Exp BinOp Exp
         | EseqExp Stm Exp
         deriving (Show, Eq)

data Stm = CompoundStm Stm Stm
         | AssignStm Id Exp
         | PrintStm [Exp]
         deriving (Show, Eq)

data St = St
  { symbolTable :: [(Id, Int)]
  , output      :: [Int]
  } deriving (Show, Eq)

prog :: Stm
prog = CompoundStm (AssignStm "a" (OpExp (NumExp 5) Plus (NumExp 3)))
                   (CompoundStm (AssignStm "b"
                                           (EseqExp (PrintStm [IdExp "a", OpExp (IdExp "a") Minus (NumExp 1)])
                                                    (OpExp (NumExp 10) Times (IdExp "a"))))
                                (PrintStm [IdExp "b"]))

maxargs :: Stm -> Int
maxargs (CompoundStm sl sr) = maxargs sl + maxargs sr
maxargs (AssignStm _ e)     = maxargsE e
maxargs (PrintStm es)       = foldl (\m e -> max m (maxargsE e)) (length es) es

maxargsE :: Exp -> Int
maxargsE (EseqExp s e) = max (maxargs s) (maxargsE e)
maxargsE _ = 0

evalS :: Stm -> St -> St
evalS (CompoundStm sl sr) st = (evalS sr . evalS sl) st
evalS (AssignStm label e) st =
  let (st', v) = evalE e st
  in  st' { symbolTable = (label, v) : symbolTable st' }
evalS (PrintStm es) st =
  foldl appendOut st es
  where
    appendOut :: St -> Exp -> St
    appendOut st' e =
      let out = output st
          (st'', v) = evalE e st'
      in  st'' { output = v : out }

evalE :: Exp -> St -> (St, Int)
evalE (IdExp l) st     = (st, fromJust $ snd <$> find ((== l) . fst) (symbolTable st))
evalE (NumExp n) st    = (st, n)
evalE (OpExp el op er) st =
  let (st', vl)        = evalE el st
      (st'', vr)       = evalE er st'
  in  (st'', runOp op vl vr)
evalE (EseqExp s e) st = evalE e (evalS s st)

runOp :: BinOp -> Int -> Int -> Int
runOp Plus v1 v2  = v1 + v2
runOp Minus v1 v2 = v1 - v2
runOp Times v1 v2 = v1 * v2
runOp Div v1 v2   = v1 `div` v2

{- Exercise 1.1 -}

type Key = String

data Tree = Leaf
          | Branch Tree Key Tree
          deriving (Show, Eq)

empty :: Tree
empty = Leaf

insert :: Key -> Tree -> Tree
insert k Leaf = Branch Leaf k Leaf
insert k t@(Branch lt ck rt)
  | k < ck = Branch (insert k lt) ck rt
  | k > ck = Branch lt ck (insert k rt)
  | otherwise = t

member :: Key -> Tree -> Bool
member _ Leaf = False
member k (Branch lt ck rt)
  | k < ck = member k lt
  | k > ck = member k rt
  | otherwise = True

data KVTree a = KVLeaf
              | KVBranch (KVTree a) Key a (KVTree a)
              deriving (Show, Eq)

kvEmpty :: KVTree a
kvEmpty = KVLeaf

kvInsert :: Key -> a -> KVTree a -> KVTree a
kvInsert key value KVLeaf = KVBranch KVLeaf key value KVLeaf
kvInsert key value tree@(KVBranch lt k v rt)
  | key < k = KVBranch (kvInsert key value lt) k v rt
  | key > k = KVBranch lt k v (kvInsert key value rt)
  | otherwise = tree

kvLookup :: Key -> KVTree a -> Maybe a
kvLookup _ KVLeaf = Nothing
kvLookup key (KVBranch lt k v rt)
  | key < k = kvLookup key lt
  | key > k = kvLookup key rt
  | otherwise = Just v
